designs/multiple-worker-node-groups.md

```diff
@@ -0,0 +1,463 @@
+metadata:
+  name: eksa-test-worker-node-template-1638469395669
+  namespace: eksa-system
+spec:
+  template:
+    spec:
+    ...
+---
+apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+kind: KubeadmConfigTemplate
+metadata:
+  name: eksa-test-md-1
+  namespace: eksa-system
+spec:
+  template:
+    spec:
+      joinConfiguration:
+        pause:
+          imageRepository: public.ecr.aws/eks-distro/kubernetes/pause
+          imageTag: v1.20.7-eks-1-20-8
+        bottlerocketBootstrap:
+          imageRepository: public.ecr.aws/l0g8r8j6/bottlerocket-bootstrap
+          imageTag: v1-20-8-eks-a-v0.0.0-dev-build.579
+        nodeRegistration:
+          criSocket: /var/run/containerd/containerd.sock
+          kubeletExtraArgs:
+            cloud-provider: external
+            tls-cipher-suites: Something
+          name: '{{ ds.meta_data.hostname }}'
+	  ...
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachineDeployment
+metadata:
+  labels:
+    cluster.x-k8s.io/cluster-name: eksa-test
+  name: eksa-test-md-1
+  namespace: eksa-system
+spec:
+  clusterName: eksa-test
+  replicas: 3
+  selector:
+    matchLabels: {}
+  template:
+    metadata:
+      labels:
+        cluster.x-k8s.io/cluster-name: eksa-test
+    spec:
+      bootstrap:
+        configRef:
+          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+          kind: KubeadmConfigTemplate
+          name: eksa-test-md-1
+      clusterName: eksa-test
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+        kind: VSphereMachineTemplate
+        name: test
+      version: v1.20.7-eks-1-20-8
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+  name: test
+  namespace: eksa-system
+spec:
+  template:
+    spec:
+    ...
+---
+```
+
+For each worker node groups, capi spec file will continue to have the following 3 kind fields.
+
+* KubeadmConfigTemplate
+* Machime deployment
+* VSphereMachineTemplate
+
+For each group, we will append these three fields corresponding to that group in the capi spec.
+
+Right now, the cli assumes that there will be only one group and it treats worker node group configuration array as a collection of only one element. As a result, the controller just refers to the first element of this array in different places of the code. So we need to do the same operations in loops, which includes capi spec creation, cluster spec validation etc. Once a capi spec is created with this approach, the workload cluster will be created with multiple worker nodes.
+
+The examples in this design are for vsphere provider. But the same strategy applies for other providers as well.
+
+## Testing
+
+To make sure, the implementation of this feature is correct, we need to add unit tests for each providers to make sure, the capi specs are generated properly.
+Also, we need to add e2e tests for each providers to test the following scenarios.
+
+* Cluster creation with one worker node group
```

Comment:
I think it's good to mention this one, for completeness. But don't we already have this kind of test?

Reply:
Yes, I just added to make sure that this doc looks complete.